yew-functional-macro/src/lib.rs

```diff
@@ -0,0 +1,175 @@
+
+                // Check here so we don't compute anything if params are invalid
+                // `>0` because first one is already consumed.
+                if inputs.len() > 0 {
+                    let params: TokenStream = inputs.map(|it| it.to_token_stream()).collect();
+                    return Err(syn::Error::new_spanned(
+                        params,
+                        "functional components can accept at most one parameter for the props",
+                    ));
+                }
+
+                let ty = match &arg {
+                    FnArg::Typed(arg) => match &*arg.ty {
+                        Type::Reference(ty) => {
+                            if ty.lifetime.is_some() {
+                                return Err(syn::Error::new_spanned(
+                                    &ty.lifetime,
+                                    "reference must not have life time",
+                                ));
+                            }
+
+                            if ty.mutability.is_some() {
+                                return Err(syn::Error::new_spanned(
+                                    &ty.mutability,
+                                    "reference must not be mutable",
+                                ));
+                            }
+
+                            ty.elem.clone()
+                        }
+                        ty => {
+                            let msg = format!(
+                                "expected a reference to a `Properties` type (try: `&{}`)",
+                                ty.to_token_stream()
+                            );
+                            return Err(syn::Error::new_spanned(ty, msg));
+                        }
+                    },
+
+                    FnArg::Receiver(_) => {
+                        return Err(syn::Error::new_spanned(
+                            arg,
+                            "functional components can't accept a receiver",
+                        ));
+                    }
+                };
+
+                Ok(Self {
+                    body: block,
+                    props_type: ty,
+                    arg,
+                    vis,
+                })
+            }
+            _ => Err(syn::Error::new(
+                Span::call_site(),
+                "`functional_component` attribute can only be applied to functions",
+            )),
+        }
+    }
+}
+
+struct FunctionalComponentName {
+    function_name: Ident,
+    component_name: Ident,
+}
+
+impl Parse for FunctionalComponentName {
+    fn parse(input: ParseStream) -> syn::Result<Self> {
+        let component_name = input.parse().map_err(|e| {
+            syn::Error::new(
+                e.span(),
+                format!("invalid name for component provided ({})", e.to_string()),
+            )
+        })?;
+        let function_name = format_ident!("Function{}", component_name);
```

Comment:
In your opinion, is this `FunctionXYZ` struct something that the user should be allowed to access (i.e. something that's part of the "official API")?

Reply:
I don't think we (will) have much of a choice here. Consider this code:
```rust
struct XYZ;

pub type ABC = XYZ;
```
This results in warning:
```
warning: private type `XYZ` in public interface (error E0446)
 --> src/main.rs:3:1
  |
3 | pub type ABC = XYZ;
  | ^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(private_in_public)]` on by default
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: for more information, see issue #34537 <https://github.com/rust-lang/rust/issues/34537>
```

This already is a warning and will become an error so I don't think we should make a part of "unofficial" or "private" API.

[Playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=8fec43abbb91330f6ecadb23949b3ad0)

Reply:
That's not what I was asking though. This isn't a question of whether we can make the struct itself private but rather whether we "define" this behaviour (i.e. document it and tell the user about it). The alternative is that we treat it as an implementation detail by giving the ident a "mixed site" span (which should make it so the user code can't access it), moving it into a `mod __impl` to discourage the user from using it, or just giving it some obscure name.

Reply:
That trait does not provide any functionality useful to the future. Unless that changes, I don't think there's a reason to expose it. The only downside I can think of is, when IDE support for proc_macros is good, it'll be a useless item in the autocomplete suggestions list.

How about we make it private right now and change it if there's a need for it?

Reply:
Yeah, that's what I had in mind too. Sadly "def site" spans are still nightly only. By giving the function provider ident `Span::def_site()` it would've been completely inaccessible from user code but oh well...

I do have a pretty good idea though! Rust allows us to declare a struct and a function with the same name (because a function is never a type and a struct isn't a function so they never overlap). So let's just give the `FunctionProvider` struct the same ident as the function has. The benefit of doing it this way is that we don't need to worry about name conflicts (if the user already has a `FunctionXYZ` type in scope the current approach errors), the identifier is essentially invisible, and if we do ever decide to expose it as part of the public API we already have a well defined and obvious naming scheme.

This would look like this: 
```rust
#[doc(hidden)]
#[allow(non_camel_case_types)]
#vis struct #fn_ident;
```
where `fn_ident` is taken from `sig.ident`.

The only thing this changes is the error message for the case where the component name is the same as the function name which becomes: "the name `xyz` is defined multiple times". We can easily handle this case by emitting a custom error for this though.

Reply:
I'd prefer just giving it an obscure name or moving it to a private module with `#[doc(hidden)]`. Using the function name to hide it just feels wrong.

Reply:
I'm sure you already know that your reasoning here isn't valid :smile:. The problem with the other approaches is that error messages (or any message that contains the type) is gonna look bad and even worse, confusing.
By giving it the same ident as the function itself it's trivial to tell which function we're talking about.

Reply:
I've made it use the function name but the error message is suboptimal when the function name and component are same. How do you think that should be handled?

Reply:
You can just compare `name == component_name` and if they are the same, emit an error pointing at the component name.
The problem you might run into is how to actually emit an error from there.

I would move the code generation code to a separate function which returns `syn::Result<TokenStream>` and then you can call `.to_compile_error()` on the `syn::Error` to turn it into a `TokenStream` (proc macro errors are literally just invocations of the `compile_error!` macro):
```rust
fn implement_functional_component(name: FunctionalComponentName, component: FunctionalComponent) -> syn::Result<TokenStream> {
    let FunctionalComponentName { component_name } = name;
    let FunctionalComponent {
        block,
        props_type,
        arg,
        vis,
        attrs,
        name,
        return_type,
    } = component;

    if component_name == name {
        return Err(syn::Error::new_spanned(component_name, "component must not have the same name as the function"));
    }
    
    let ret_type = quote_spanned!(return_type.span() => ::yew::html::Html);

    quote! {
        #[doc(hidden)]
        #[allow(non_camel_case_types)]
        #vis struct #name;

        impl ::yew_functional::FunctionProvider for #name {
            type TProps = #props_type;

            fn run(#arg) -> #ret_type {
                #block
            }
        }

        #(#attrs)*
        #vis type #component_name = ::yew_functional::FunctionComponent<#name>;
    }
}

#[proc_macro_attribute]
pub fn functional_component(
    attr: proc_macro::TokenStream,
    item: proc_macro::TokenStream,
) -> proc_macro::TokenStream {
    let name = parse_macro_input!(attr as FunctionalComponentName);
    let component = parse_macro_input!(item as FunctionalComponent);
    
    implement_functional_component(name, component)
        .unwrap_or_else(|err| err.to_compile_error())
        .into()
}
```